Automattic/sensei
# Optimize JS script file loading

Fixes #1530.

In investigating the screens that Select2 should be loaded on, I realized that `lesson-metadata.min.js` was being loaded on pages that didn't use it. This file contains JS specific to lessons and questions and [loads Select2 as a dependency](https://github.com/Automattic/sensei/blob/08c542118523f2a1e726fdccb5f32ff60c833111/includes/class-sensei-lesson.php#L2138). Therefore, part of the work for this PR involves only loading `lesson-metadata.min.js` when adding or editing lessons or questions, and then loading Select2 separately on pages that need it.

It was a similar situation for `modules-admin.min.js`, which was also being loaded on pages that didn't use it.

This PR also moves course and lesson ordering logic from `settings.js` to a new `ordering.js` file that is only loaded on the _Order Courses_ and _Order Lessons_ pages.

_Note that this PR does not handle Select2 CSS, which is still being loaded on every page in WP admin._

## Testing Instructions
- Check that ordering courses and lessons work.
- Check that you can edit the start date and add a learner to a course in _Sensei LMS_ > _Learner Management_ > _Manage Learners_.
- Check that drop-downs are converted to Select2 drop-downs on the following pages:
  - Grading
  - Learner Management
  - Add / Edit Course
  - Add Module / Edit Module (only the Course(s) field)
  - Order Modules
  - Add / Edit Lesson
  - Order Lessons
  - Add Question
- With DevTools open, check that `select2.full.min.js` is only loaded on the following Sensei LMS pages:
  - Grading
  - Learner Management
  - Add / Edit Course
  - Order Courses
  - Modules / Edit Module
  - Order Modules
  - Add / Edit Lesson
  - Order Lessons
  - Add / Edit Question
- With DevTools open, check that `modules-admin.min.js` is only loaded on the following Sensei LMS pages:
  - Modules
  - Edit Module
  - Order Modules
- With DevTools open, check that `lesson-metadata.min.js` is only loaded on the following Sensei LMS pages:
  - Add / Edit Lesson
  - Add / Edit Question
